package/MDAnalysis/analysis/waterdynamics.py

```diff
@@ -1194,99 +1194,108 @@ class SurvivalProbability(object):
-        """
-        This function gets one point of the plot P(t) vs t. It uses the
-        _getOneDeltaPoint() function to calculate the average.
-        """
-        n = 0
-        sumDeltaP = 0.0
-        for frame_no in range(len(selected) - window_size):
-            delta = self._getOneDeltaPoint(selected, frame_no, window_size)
-            sumDeltaP += delta
-            n += 1
+        if tau_max > (stop - start):
+            raise ValueError("Too few frames selected for given tau_max.")
 
-        return sumDeltaP/n
+        # load all frames to an array of sets
+        selected_ids = []
+        for ts in self.universe.trajectory[start:stop]:
+            self.print(verbose, "Loading frame:", ts)
+            selected_ids.append(set(self.universe.select_atoms(self.selection).ids))
 
+        tau_timeseries = np.arange(1, tau_max + 1)
+        sp_timeseries = [[] for _ in range(tau_max)]
 
-    def _getOneDeltaPoint(self, selected, t, tau):
-        """
-        Gives one point to calculate the mean and
-        gets one point of the plot C_vect vs t.
-        - Ex: t=1 and tau=1 calculates
-        how many selected water molecules survive from the frame 1 to 2
-        - Ex: t=5 and tau=3 calculates
-        how many selected water molecules survive from the frame 5 to 8
-        """
+        for t in range(0, len(selected_ids), step):
+            Nt = len(selected_ids[t])
 
-        Nt = len(selected[t])
-        if Nt == 0:
-            return 0
+            if Nt == 0:
+                self.print(verbose,
+                           "At frame {} the selection did not find any molecule. Moving on to the next frame".format(t))
+                continue
 
-        # fraction of water molecules that survived
-        Ntau = self._NumPart_tau(selected, t, tau)
-        return Ntau/Nt
+            for tau in tau_timeseries:
+                if t + tau >= len(selected_ids):
+                    break
 
+                # ids that survive from t to t + tau and at every frame in between
+                Ntau = len(set.intersection(*selected_ids[t:t + tau + 1]))
+                sp_timeseries[tau - 1].append(Ntau / float(Nt))
 
-    def _NumPart_tau(self, selected, t, tau):
-        """
-        Compares the molecules in t selection and t+tau selection and
-        select only the particles that remain from t to t+tau and
-        at each point in between.
-        It returns the number of remaining particles.
-        """
-        survivors = set(selected[t])
-        i = 0
-        while (t + i) < t + tau and (t + i) < len(selected):
-            next = set(selected[t + i])
-            survivors = survivors.intersection(next)
-            i += 1
-        return len(survivors)
\ No newline at end of file
+        # user can investigate the distribution and sample size
+        self.sp_timeseries = sp_timeseries
+
+        return tau_timeseries, [np.mean(sp) for sp in sp_timeseries]
```

Comment:
It's better here to save the attributes to the class and return the class from `run` (`return self`)

Reply:
*better* means that it follows the behavior of other analysis classes in MDAnalysis.  The general pattern to use *any* analysis class in MDAnalysis is this.

```python
sp = SurvivalProbability(....).run()
sp.tau
sp.means
```

If we keep all classes to follow this it will make MDAnalysis easier/more intuitive to use.

Reply:
Thanks @richardjgowers and @kain88-de 

I've updated the code to conform to your comments. 

If you don't mind I will this opportunity to share some thoughts. 1) @kain88-de The constructor in python has to return None, 2) I have not seen a single API that returns data in the object's field, I imagine that you have some bigger picture in mind that I missing,  3) If the purpose of returning self is to be able to use again .run() straight away, that would overwrite the previous data. 

I'll update the comments/documentation once all the code is reviewed. Thanks

Reply:
1 ) I'm not sure what you mean. But yeah the `__ini__` function doesn't return anything. But `Object()` does return an instance of the class after `__init__` was called.
2 ) See `rmsd.RMSD`, `rmsd.RMSF`, `contacts.Contacts`. and other analysis classes using the base analysis class store results in `self`. 
3 )  It allows writing a [base class](https://github.com/MDAnalysis/mdanalysis/blob/461aadf6733d234e0913bae774a6312a208ede27/package/MDAnalysis/analysis/base.py#L44) that shares reusable code between different analysis types. Having `run` return itself allows method chaining and code I've written above. Mainly it fits with how the rest of analysis classes work. Consistency in API is key here.

Reply:
Oh, of course you're right about 1). Sorry, my mistake. Thanks, I'll have a look at the base class too. Let me know if you see anything else in the code that needs attention. Mat